share/p2p/discovery/discovery.go

```diff
@@ -202,21 +204,32 @@ func (d *Discovery) Advertise(ctx context.Context) {
 // It initiates peer discovery upon request and restarts the process until the soft limit is
 // reached.
 func (d *Discovery) discoveryLoop(ctx context.Context) {
-	t := time.NewTicker(discoveryRetryTimeout)
-	defer t.Stop()
+	backoffDuration := backoffInitialDuration
+	backoffTicker := time.NewTicker(backoffDuration)
+	defer backoffTicker.Stop()
 
 	warnTicker := time.NewTicker(logInterval)
 	defer warnTicker.Stop()
 
+	// Function to initiate discovery and restart the timer
+	runDiscovery := func() {
+		if !d.discover(ctx) {
+			// rerun discovery if the number of peers hasn't reached the limit
+			return
+		}
+
+		backoffTicker.Stop()
+		backoffDuration = backoffInitialDuration
+		backoffTicker = time.NewTicker(backoffDuration)
+	}
+
 	for {
-		// drain all previous ticks from the channel
-		drainChannel(t.C)
 		select {
-		case <-t.C:
-			if !d.discover(ctx) {
-				// rerun discovery if the number of peers hasn't reached the limit
-				continue
+		case <-backoffTicker.C:
+			if backoffDuration > maxBackoffDuration {
+				backoffDuration = maxBackoffDuration
```

Comment:
would be better to have this logic near the place where `backoffDuration` is updated